azure db discovery helm chart (#16867)
gravitational/teleport: docs/pages/reference/helm-reference/teleport-kube-agent.mdx

      regions: ["eu-west-1"]
      tags:
        "*": "*"
  annotations:
    serviceAccount:
      eks.amazonaws.com/role-arn: arn:aws:iam::1234567890:role/my-rds-autodiscovery-role
  ```

  <Admonition type="note" title="YAML formatting">
    YAML is very sensitive to correct spacing. When specifying lists in a `values.yaml` file, you might like
    to [use a linter](https://codebeautify.org/yaml-validator) to validate your YAML list and ensure that it is correctly formatted.
  </Admonition>
  (!docs/pages/includes/yaml-lint-note.mdx!)
  </TabItem>

>   <TabItem label="--set">

Reviewer: Why remove the `--set` version of this and the other edited sections?

Author: The `--set` tab was extremely cumbersome to use for config - it seems misleading to even suggest someone use that method for configuration. Hugo suggested removing it and I agreed that no one is realistically going to bother with --set here. I don't care if we bring it back though, so it's up to you

>   This section configures Azure database auto-discovery. You can configure databases for other platforms using the [`awsDatabases`](#awsDatabases) or [`databases`](#databases) sections.

Reviewer: ```suggestion
  This section configures Azure database auto-discovery. You can configure databases for other platforms using the [`awsDatabases`](#awsdatabases) or [`databases`](#databases) sections.
```

  First, create in the chart installation namespace a Kubernetes `Secret` containing the azure client secret:
  ```code
  $ kubectl create secret generic teleport-azure-client-secret --from-literal=client_secret=<your-azure-client-secret>
  secret/teleport-azure-client-secret created
  ```

  Then, use the [`extraEnv`](#extraenv) value to set the pods environment variables:

  ```yaml
  extraEnv:
  - name: AZURE_CLIENT_SECRET
    valueFrom:
      secretKeyRef:
      name: teleport-azure-client-secret
      key: client_secret
      optional: false
  - name: AZURE_TENANT_ID
    value: "11111111-2222-3333-4444-555555555555"
  - name: AZURE_CLIENT_ID
    value: "11111111-2222-3333-4444-555555555555"
  ```
</Admonition>

| Type | Default value | Required? |
| - | - | - |
| `list` | `[]` | When the `db` chart role is used at least one of `databases`, `awsDatabases`, `azureDatabases`, `databaseResources` is required. |

`azureDatabases` is a YAML list object detailing the filters for the Azure databases that should be discovered and proxied by Teleport Database access.

You can specify multiple database filters by adding additional list elements.

Required fields for each filter:

> - `types` is a list containing the types of Azure databases that should be discovered.

Author: `[array]` - will change

> - `types` is a list containing the types of Azure databases that should be discovered.

Author: Oh, actually we show the type to be `list` above and the default value is `[]`. We do the same throughout this doc. I think list is clear enough.

  </Admonition>

  </TabItem>
</Tabs>

## `azureDatabases`

<Admonition type="tip" title="Azure database auto-discovery">
  This section configures Azure database auto-discovery. You can configure databases for other platforms using the [`awsDatabases`](#awsDatabases) or [`databases`](#databases) sections.
</Admonition>

<Admonition type="note" title="Azure IAM">
  For Azure database auto-discovery to work, your Database Service pods will need to have appropriate IAM permissions as per the [database documentation](../../database-access/guides/azure-postgres-mysql.mdx#step-35-configure-iam-permissions-for-teleport).

  After configuring a service principal with appropriate IAM permissions, you can use
  `extraEnv` to provide credentials as environment variables to your pods:

  ```
  extraEnv:
  - name: AZURE_TENANT_ID
    value: "11111111-2222-3333-4444-555555555555"
  - name: AZURE_CLIENT_ID
    value: "11111111-2222-3333-4444-555555555555"
  - name: AZURE_CLIENT_SECRET
    value: "123456789abcdefghijklmnopqrstuvwxyzABCDE"
  ```
</Admonition>

| Type | Default value | Required? |
| - | - | - |

> | `list` | `[]` | When the `db` chart role is used at least one of `databases`, `awsDatabases`, `azureDatabases`, `dbResources` is required. |

Author: Will update. Also, I think `dbResources` is incorrect - shouldn't it be `databaseResources`?

> | `list` | `[]` | When the `db` chart role is used at least one of `databases`, `awsDatabases`, `azureDatabases`, `dbResources` is required. |

Reviewer 2: Yup, I renamed it at some point and forgot some mentions, the correct one is `databaseResources`

> | `list` | `[]` | When the `db` chart role is used at least one of `databases`, `awsDatabases`, `azureDatabases`, `dbResources` is required. |

Author: will update then

`azureDatabases` is a YAML list object detailing the filters for the Azure databases that should be discovered and proxied by Teleport Database access.

You can specify multiple database filters by adding additional list elements.

Required fields for each filter:
- `types` is a list containing the types of Azure databases that should be discovered.
- `tags` can be used to set Azure resource tags that must be matched for databases to be discovered.

Optional fields for each filter:
- `regions` is a list of Azure regions which should be scanned for databases.
- `subscriptions` can be used to discover databases within matching Azure subscriptions.
- `resource_groups` can be used to discover databses within matching Azure resource groups.

> 

Reviewer 2: ```suggestion

The default for each of these optional settings is `*`, which will auto-discover in all
subscriptions, regions, or resource groups accessible by the Teleport service
principal in Azure.

```

<Tabs>
  <TabItem label="values.yaml">
  ```yaml
  azureDatabases:
  - types: ["mysql", "postgres"]
    tags:
      "*": "*"
  - types: ["mysql"]
    tags:
      "env": ["dev", "staging"]
      "origin": "alice"
    regions: ["eastus", "centralus"]
    subscriptions: ["subID1", "subID2"]
    resource_groups: ["group1", "group2"]

>   ```

Reviewer 2: I think we should add `roles: db` here as it is required for the feature to work. Maybe `extraEnv` also.

I don't think you need to bother with the `--set` tab, using azure database requires too many arguments, this is error-prone and I doubt anyone will do it.

Author: Yeah I noticed that we didn't include all required fields in our examples, e.g. proxyAddr.

I'll provide a full working example config with all required fields.

> I don't think you need to bother with the --set tab, using azure database requires too many arguments, this is error-prone and I doubt anyone will do it.

Agreed, I think it would be easier to provide a shell here-doc yaml example than using the very limited --set syntax, or just provide only the yaml file example.

I provided --set only because that was the convention for other examples, but if we're ok with breaking convention I'll remove it (I really doubt anyone wants to use --set as you mentioned).

`azureDatabases` is a YAML list object detailing the filters for the Azure databases that should be discovered and proxied by Teleport Database access.

You can specify multiple database filters by adding additional list elements.

Required fields for each filter:
- `types` is a list containing the types of Azure databases that should be discovered.
- `tags` can be used to set Azure resource tags that must be matched for databases to be discovered.

Optional fields for each filter:
- `regions` is a list of Azure regions which should be scanned for databases.
- `subscriptions` can be used to discover databases within matching Azure subscriptions.
- `resource_groups` can be used to discover databses within matching Azure resource groups.

<Tabs>
  <TabItem label="values.yaml">
  ```yaml
  azureDatabases:
  - types: ["mysql", "postgres"]
    tags:
      "*": "*"
  - types: ["mysql"]
    tags:
      "env": ["dev", "staging"]
      "origin": "alice"
    regions: ["eastus", "centralus"]
    subscriptions: ["subID1", "subID2"]
    resource_groups: ["group1", "group2"]
  ```

  <Admonition type="note" title="YAML formatting">

>     YAML is very sensitive to correct spacing. When specifying lists in a `values.yaml` file, you might like
>     to [use a linter](https://codebeautify.org/yaml-validator) to validate your YAML list and ensure that it is correctly formatted.

Reviewer 3: Nit: for some reason I can't articulate this sounds weird to me. Take it or leave it:
```suggestion
    YAML is very sensitive to correct spacing. When specifying lists in a `values.yaml` file, you might find it helpful
    to [use a linter](https://codebeautify.org/yaml-validator) to validate your YAML list and ensure that it is correctly formatted.
```

Author: Sounds good, I actually copied this from the above awsDatabases doc, so I'll change that too

Author: oh it's actually in several places, I'll make it an include snippet

  ```
</Admonition>

## `databaseResources`

| Type | Default value | Required? |
| - | - | - |

> | `list` | `[]` | When the `db` chart role is used at least one of `databases`, `awsDatabases`, `azureDatabases`, `databaseResources` is required. |

Reviewer: As noted above:

```suggestion
| `list` | `[]` | When the `db` chart role is used at least one of `databases`, `awsDatabases`,<br/> `azureDatabases`, `databaseResources` is required. |
```
